controllers/medusa/medusatask_controller.go

```diff
@@ -0,0 +1,470 @@
+		task.Status.FinishTime = metav1.Now()
+		if err := r.Status().Patch(ctx, task, patch); err != nil {
+			logger.Error(err, "failed to patch status with finish time")
+			return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+		}
+
+		// Schedule a sync if the task is a purge
+		if task.Spec.Operation == medusav1alpha1.OperationTypePurge {
+			r.scheduleSyncForPurge(task)
+		}
+
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	// If the task is already finished, there is nothing to do.
+	if taskFinished(task) {
+		logger.Info("Task is already finished", "MedusaTask", req.NamespacedName, "Operation", task.Spec.Operation)
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	// Task hasn't started yet
+	cassdcKey := types.NamespacedName{Namespace: task.Namespace, Name: task.Spec.CassandraDatacenter}
+	cassdc := &cassdcapi.CassandraDatacenter{}
+	err = r.Get(ctx, cassdcKey, cassdc)
+	if err != nil {
+		logger.Error(err, "failed to get cassandradatacenter", "CassandraDatacenter", cassdcKey)
+		return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+	}
+
+	pods, err := r.getCassandraDatacenterPods(ctx, cassdc, logger)
+	if err != nil {
+		logger.Error(err, "Failed to get datacenter pods")
+		return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+	}
+
+	if task.Spec.Operation == medusav1alpha1.OperationTypePurge {
+		return r.purgeOperation(ctx, task, pods, logger)
+	} else if task.Spec.Operation == medusav1alpha1.OperationTypeSync {
+		return r.syncOperation(ctx, task, pods, logger)
+	} else if task.Spec.Operation == medusav1alpha1.OperationTypePrepareRestore {
+		return r.prepareRestoreOperation(ctx, task, pods, logger)
+	} else {
+		return ctrl.Result{}, fmt.Errorf("unsupported operation %s", task.Spec.Operation)
+	}
+}
+
+func (r *MedusaTaskReconciler) getCassandraDatacenterPods(ctx context.Context, cassdc *cassdcapi.CassandraDatacenter, logger logr.Logger) ([]corev1.Pod, error) {
+	podList := &corev1.PodList{}
+	labels := client.MatchingLabels{cassdcapi.DatacenterLabel: cassdc.Name}
+	if err := r.List(ctx, podList, labels); err != nil {
+		logger.Error(err, "failed to get pods for cassandradatacenter", "CassandraDatacenter", cassdc.Name)
+		return nil, err
+	}
+
+	pods := make([]corev1.Pod, 0)
+	pods = append(pods, podList.Items...)
+
+	return pods, nil
+}
+
+func (r *MedusaTaskReconciler) purgeOperation(ctx context.Context, task *medusav1alpha1.MedusaTask, pods []corev1.Pod, logger logr.Logger) (reconcile.Result, error) {
+	logger.Info("Starting purge operations")
+
+	// Set the start time and add the pods to the in progress list
+	patch := client.MergeFrom(task.DeepCopy())
+	task.Status.StartTime = metav1.Now()
+	for _, pod := range pods {
+		task.Status.InProgress = append(task.Status.InProgress, pod.Name)
+	}
+
+	if err := r.Status().Patch(ctx, task, patch); err != nil {
+		logger.Error(err, "Failed to patch status")
+		// We received a stale object, requeue for next processing
+		return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
```

Comment:
No need to set `RequeueAfter` when returning a non-nil error.

Reply:
👍